src/wire/chunked_report.rs

```diff
@@ -0,0 +1,469 @@
+    pub fn is_last_chunk(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::IS_LAST_CHUNK]
+    }
+
+    /// Return the `reserved` field
+    #[inline]
+    pub fn reserved(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::RESERVED]
+    }
+
+    /// Return the `n_chunk_payload_bytes` field
+    #[inline]
+    pub fn n_chunk_payload_bytes(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::N_CHUNK_PAYLOAD_BYTES]
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> ChunkedReport<&'a T> {
+    /// Return a pointer to the payload
+    #[inline]
+    pub fn payload(&self) -> &'a [u8] {
+        let data = self.buffer.as_ref();
+        &data[field::PAYLOAD]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> ChunkedReport<T> {
+    /// Set the `fingerprint` field
+    #[inline]
+    pub fn set_fingerprint(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::FINGERPRINT], value);
+    }
+
+    /// Set the `probe_id` field
+    #[inline]
+    pub fn set_probe_id(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::PROBE_ID], value);
+    }
+
+    /// Set the `chunk_group_id` field
+    #[inline]
+    pub fn set_chunk_group_id(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u16(&mut data[field::CHUNK_GROUP_ID], value);
+    }
+
+    /// Set the `chunk_index` field
+    #[inline]
+    pub fn set_chunk_index(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u16(&mut data[field::CHUNK_INDEX], value);
+    }
+
+    /// Set the `payload_data_type` field
+    #[inline]
+    pub fn set_payload_data_type(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::PAYLOAD_DATA_TYPE] = value;
+    }
+
+    /// Set the `is_last_chunk` field
+    #[inline]
+    pub fn set_is_last_chunk(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::IS_LAST_CHUNK] = value;
+    }
+
+    /// Set the `reserved` field
+    #[inline]
+    pub fn set_reserved(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::RESERVED] = value;
+    }
+
+    /// Set the `n_chunk_payload_bytes` field
+    #[inline]
+    pub fn set_n_chunk_payload_bytes(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::N_CHUNK_PAYLOAD_BYTES] = value;
+    }
+
+    /// Return a mutable pointer to the payload
+    #[inline]
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        let data = self.buffer.as_mut();
+        &mut data[field::PAYLOAD]
+    }
+}
+
+impl<T: AsRef<[u8]>> AsRef<[u8]> for ChunkedReport<T> {
+    fn as_ref(&self) -> &[u8] {
+        self.buffer.as_ref()
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> AsMut<[u8]> for ChunkedReport<T> {
+    fn as_mut(&mut self) -> &mut [u8] {
```

Comment:
Similar query as before -- are we sure we need `as_mut`? I think it would be safer to require external users that want to muck with the underlying bytes to deconstruct/unwrap this instance then re-wrap it (optionally with checking) after they're done.

Reply:
Yeah I think your right.
I thought it be a convenience thing, but after thinking about it again, seems more like a :foot: :gun:.
I'll get those removed.